plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts

```diff
@@ -99,106 +100,98 @@ export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
         }),
       );
     }
-    this.running = true;
-    this.run();
-  }
 
-  private async run() {
-    while (this.running) {
-      try {
-        // TODO: We want to disconnect the queue popping and message processing
-        // so that if the queue popping fails we exponentially back off in order to give the DB room to sort itself out.
-        await this.process();
-      } catch (e) {
-        this.logger.warn('Processing failed with:', e);
-        // TODO: this can be a little smarter as mentioned in the above comment.
-        // But for now, if something fails, wait a brief time to pick up the next message.
-        await this.wait();
-      }
+    if (this.stopFunc) {
+      throw new Error('Processing engine is already started');
     }
-  }
 
-  private async process() {
-    const { items } = await this.processingDatabase.transaction(async tx => {
-      return this.processingDatabase.getProcessableEntities(tx, {
-        processBatchSize: 1,
-      });
-    });
+    this.stopFunc = startTaskPipeline<RefreshStateItem>({
+      lowWatermark: 5,
+      highWatermark: 10,
+      loadTasks: async count => {
+        try {
+          const { items } = await this.processingDatabase.transaction(
+            async tx => {
+              return this.processingDatabase.getProcessableEntities(tx, {
+                processBatchSize: count,
+              });
+            },
+          );
+          return items;
+        } catch (error) {
```

Comment:
Could perhaps be argued that the pipeline should handle thrown errors (it'll have to, anyway, and may be the right place to handle exponential backoff and stuff

Reply:
I experimented with an error handler that you passed to the pipeline, but it got real weird when having to deal with errors from both the loader and the processor, so ended up with this design instead where the callbacks have to make sure they don't throw anything. In practice it does end up meaning that the loader falls back to polling if it's completely down as it'll end up not being able to refill work items. Exponential backoff is something is something we could add if this ends up being reused elsewhere where it makes more sense.